### PR TITLE
Allow map to be called for arbitrary depth

### PR DESCRIPTION
**Pull request by dsmilkov**
Map can now be called for any level of repeated. I need this when doing clustering so I can call it for any `conversations.*.from`

Also:
- Fix some deps in pyproject (umap should be optional)

### CODE REVIEW

File: lilac/batch_utils.py [hunk 1]
```diff
@@ -6,54 +6,45 @@
...
   original_input: Union[Iterable, object],
-  is_primitive_predicate: Callable[[object], bool] = is_primitive,
+  max_depth: int = -1,
 ) -> Generator:
   """Unflattens a deeply flattened iterable according to the original iterable's structure."""
   flat_input_iter = iter(flat_input)
-  if isinstance(original_input, Iterable) and not is_primitive_predicate(original_input):
+  if isinstance(original_input, Iterable) and not is_primitive(original_input):
     for o in original_input:
-      yield _array_unflatten(flat_input_iter, o, is_primitive_predicate)
```
Comment (nsthorat) [inline lilac/batch_utils.py:53]: do we use is primitive predicate anywhere now?

File: lilac/batch_utils.py [hunk 2]
```diff
@@ -6,54 +6,45 @@
...
 
 def _array_unflatten(
-  flat_input: Iterator[list[object]],
-  original_input: Union[Iterable, object],
-  is_primitive_predicate: Callable[[object], bool],
+  flat_input: Iterator[list[object]], original_input: Union[Iterable, object], depth: int
 ) -> Union[list, dict]:
   """Unflattens a deeply flattened iterable according to the original iterable's structure."""
-  if is_primitive_predicate(original_input) or isinstance(original_input, dict):
+  if depth == 0 or is_primitive(original_input) or isinstance(original_input, dict):
```
Comment (nsthorat) [inline lilac/batch_utils.py:28]: s/depth/max_depth for consistency

File: lilac/batch_utils.py [hunk 3]
```diff
@@ -6,54 +6,45 @@
...
-def array_flatten(
-  input: Union[Iterator, Iterable], is_primitive_predicate: Callable[[object], bool] = is_primitive
-) -> Iterator:
-  """Flattens a deeply nested iterator.
-
-  Primitives and dictionaries are not flattened. The user can also provide a predicate to determine
-  what is a primitive.
-  """
-  return _array_flatten(input, is_primitive_predicate)
+def array_flatten(input: Union[Iterator, Iterable], max_depth: int = -1) -> Iterator:
```
Comment (nsthorat) [inline lilac/batch_utils.py:18]: actually this might be better as iterable_flatten since it's not a list

File: lilac/data/dataset_utils.py [hunk 4]
```diff
@@ -65,7 +65,7 @@ def count_primitives(input: Union[Iterable, Iterator]) -> int:
 
   Sum the final set of counts. This is the important iterable not to exhaust.
   """
-  return sum((len(list(array_flatten(i))) for i in input))
+  return sum(1 for _ in array_flatten(input))
```
Comment (nsthorat) [inline lilac/data/dataset_utils.py:68]: len(list(array_flatten))?

File: lilac/data/dataset_utils.py [hunk 5]
```diff
@@ -65,7 +65,7 @@ def count_primitives(input: Union[Iterable, Iterator]) -> int:
```
Comment (nsthorat) [inline lilac/data/dataset_utils.py:65]: This seems like it's just for counting inputs == outputs, so maybe we just call it count_leafs or something

**Review by nsthorat — APPROVED** [review]

File: lilac/batch_utils.py (hunk 1 above)
Comment (dsmilkov) [reply lilac/batch_utils.py:53]: nope, we haven't used it at HEAD for a while now

**Comment by dsmilkov** [comment]
all done

**Pull request merged**